src/nix/profile.cc

```diff
@@ -453,55 +453,117 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+            return element.storePaths.count(storePath);
+        }
+    };
+}
+
+Matcher createNameMatcher(const std::string & name) {
+    return {
+        .name = fmt("Package name '%s'", name),
+        .matches = [name](const std::string &elementName, const ProfileElement & element) {
+            return elementName == name;
+        }
+    };
+}
+
+Matcher all = {
+    .name = "All",
+    .matches = [](const std::string &name, const ProfileElement & element) {
+        return true;
+    }
+};
+
+class MixProfileElementMatchers : virtual Args, virtual StoreCommand
+{
+    std::vector<Matcher> _matchers;
 
 public:
 
     MixProfileElementMatchers()
     {
-        expectArgs("elements", &_matchers);
+        addFlag({
+            .longName = "all",
+            .description = "Matches all packages in profile.",
+            .handler = {[this]() {
+                _matchers.push_back(all);
+            }},
+        });
+        addFlag({
+            .longName = "regex",
+            .description = "A regular expression to match one or more packages in profile.",
+            .labels = {"pattern"},
+            .handler = {[this](std::string arg) {
+                _matchers.push_back(createRegexMatcher(arg));
+            }},
+        });
+        expectArgs({
+            .label = "elements",
+            .optional = true,
+            .handler = {[this](std::vector<std::string> args) {
+                for (auto & arg : args) {
+                    if (auto n = string2Int<size_t>(arg)) {
+                        throw Error("'nix profile' no longer supports indices ('%d')", *n);
+                    } else if (getStore()->isStorePath(arg)) {
+                        _matchers.push_back(createStorePathMatcher(getStore()->parseStorePath(arg)));
+                    } else {
+                        _matchers.push_back(createNameMatcher(arg));
+                    }
+                }
+            }}
+        });
     }
 
-    struct RegexPattern {
-        std::string pattern;
-        std::regex  reg;
-    };
-    typedef std::variant<Path, RegexPattern> Matcher;
-
-    std::vector<Matcher> getMatchers(ref<Store> store)
-    {
-        std::vector<Matcher> res;
+    std::set<std::string> getMatchingElementNames(ProfileManifest & manifest) {
+        if (_matchers.empty()) {
+            throw UsageError("No packages specified.");
+        }
 
-        for (auto & s : _matchers) {
-            if (auto n = string2Int<size_t>(s))
-                throw Error("'nix profile' no longer supports indices ('%d')", *n);
-            else if (store->isStorePath(s))
-                res.push_back(s);
-            else
-                res.push_back(RegexPattern{s,std::regex(s, std::regex::extended | std::regex::icase)});
+        if (std::find_if(_matchers.begin(), _matchers.end(), [](const Matcher & m) { return m.name == "All"; }) != _matchers.end() && _matchers.size() > 1) {
```

Comment:
Checking for `"All"` is brittle, maybe `Matcher` can have a field like `isAll`?

Reply:
I've opted to add a type to the `Matcher` to distinguish between the types of matchers.